fuzz/fuzz_lc3.cc

```diff
@@ -0,0 +1,66 @@
+#include "lc3.h"
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <fuzzer/FuzzedDataProvider.h>
+#include <vector>
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  FuzzedDataProvider fdp(data, size);
+  auto new_samplerate = [&] {
+    return fdp.PickValueInArray({8000, 16000, 24000, 32000, 48000});
+  };
+
+
+  auto new_pcm_format = [&] {
+    return fdp.PickValueInArray({LC3_PCM_FORMAT_S16, LC3_PCM_FORMAT_S24,
+                                 LC3_PCM_FORMAT_S24_3LE, LC3_PCM_FORMAT_FLOAT});
+  };
+
+  const int dt_us = fdp.PickValueInArray({7500, 10000});
+  const int sr_hz = new_samplerate();
+
+  assert(LC3_CHECK_SR_HZ(sr_hz));
+
+  int sr_pcm_hz = new_samplerate();
+  assert(LC3_CHECK_SR_HZ(sr_pcm_hz));
+
+  if (sr_pcm_hz > sr_hz) {
+    sr_pcm_hz = 0;
+  }
+  const lc3_pcm_format format = new_pcm_format();
+
+  int encoder_size = lc3_encoder_size(dt_us, sr_hz);
+  void *mem = malloc(lc3_encoder_size(dt_us, sr_hz));
+  lc3_encoder_t enc = lc3_setup_encoder(dt_us, sr_hz, sr_pcm_hz, mem);
+
+  if (enc == NULL) {
+    free(mem);
+    return -1;
+  }
+
+  alignas(alignof(uint32_t)) std::array<uint8_t, LC3_MAX_FRAME_BYTES> out = {0};
```

Comment:
Oh that makes things a little easier. I was actually hoping to allocate this dynamically, rather than an array of size LC3_MAX_FRAME_BYTES, but I couldn't work out the correct size to allocate. Is there some helper to work out the output size based on a given input size?

The reason I ask is that, when you combine the fuzzer with an address sanitizer if you allocate exactly the right size it'll detect overflows (beyond that size) using padded shadow bytes.

Reply:
There is no rule between input and output size. It's completely a free user choice.
You have to consider input and output to determine a compression ratio, but all the range can be applied for any input format.